pkg/issuer/ca/setup_test.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ca
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+
+	cmv1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	"github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/controller/certificaterequests/ca"
+	controllertest "github.com/jetstack/cert-manager/pkg/controller/test"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+	"github.com/jetstack/cert-manager/test/unit/gen"
+	"github.com/jetstack/cert-manager/test/unit/listers"
+)
+
+func TestCA_Setup(t *testing.T) {
+	caCrt, caKey := mustGenerateTLSAssets(t)
+	tests := map[string]struct {
+		givenNamespace string
+		givenSecret    *corev1.Secret
+		givenIssuer    cmv1.GenericIssuer
+		wantCert       *cmv1.Certificate
+		wantErr        error
+	}{
+		"a": {
+			givenIssuer: &cmv1.Issuer{
+				Spec: cmv1.IssuerSpec{IssuerConfig: cmv1.IssuerConfig{
+					CA: &cmv1.CAIssuer{
+						SecretName:  "secret-1",
+						OCSPServers: []string{"http://ocsp-v3.example.org"},
+					},
+				}},
+			},
+			givenSecret: gen.SecretFrom(gen.Secret("secret-1"),
+				gen.SetSecretNamespace("default"),
+				gen.SetSecretData(map[string][]byte{
+					"tls.key": caKey,
+					"tls.crt": caCrt,
+				}),
+			),
+			givenNamespace: "default",
+			wantErr:        nil,
+		},
+	}
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			rec := &controllertest.FakeRecorder{}
+
+			c := &CA{
+				Context: &controller.Context{
+					Recorder: rec,
+				},
+				issuer:            test.givenIssuer,
+				resourceNamespace: test.givenNamespace,
+				secretsLister: listers.FakeSecretListerFrom(listers.NewFakeSecretLister(),
+					listers.SetFakeSecretNamespaceListerGet(test.givenSecret, nil),
+				),
+			}
+
+			err := c.Setup(context.Background())
+			// TODO: How do I c.Issue()? The ca.CA struct only implements
+			// the Setup function, not the Issue function.
```

Comment:
> A unit test in the CA issuer code that ensures that given an issuer that configures `ocspServers` , a resulting certificate will contain those OCSP servers would suffice.

@munnerz How can I test that [ca.CA](https://github.com/jetstack/cert-manager/blob/ab0cd57dc58fd73a76fd96bd9d1402bd5ae96582/pkg/issuer/ca/ca.go#L31-L40) can issue a certificate? ca.CA does not seem to implement `Issue()` :(

Reply:
I tried writing a new unit test in `pkg/controller/certificaterequests/ca` but I had no way to check that the oscpServers has been properly set in the resulting certificate.

I started to write a new unit test that does check the created certificate; you can see it on another branch (take a look at [ca_test.go](https://github.com/maelvls/cert-manager/blob/5fcaad7a501cb582b8272672cb8fdc303ea25148/pkg/controller/certificaterequests/ca/ca_test.go#L404-L506)).

I have not finished and I really wish we could add this ocspServers feature to v1.2-alpha.2 we are going to release this morning

@munnerz What do you think? I will create a new issue "add unit test for `ocspServers` and `crlDistributionPoints`" and point to the existing branch I started